dxp/layout_rendering.md

```diff
@@ -0,0 +1,173 @@
+# Layout (page) rendering in DXP
+
+This document explains how a request to render a page is serviced in DXP.
+
+## What gets requested when asking for http://localhost:8080?
+
+First, we are going to list what external and internal requests are fired whenever someone hits DXP's root page.
+
+> Note that, in addition to the requests a browser can make to our server, the server can dispatch internal requests to resources which undergo the standard request lifecycle, simulating an external request without any need to hit the HTTP transport layer at all. Usually the responses of these internal requests are included in the originating (external) request.
+
+> All this is done leveraging the [`RequestDispatcher` API](https://docs.oracle.com/javaee/7/api/javax/servlet/RequestDispatcher.html) that can be retrieved from any [`ServletRequest`](https://javaee.github.io/javaee-spec/javadocs/javax/servlet/ServletRequest.html#getRequestDispatcher-java.lang.String-). You can see an example of how it is used in [`MainServlet.java`](https://github.com/liferay/liferay-portal/blob/2448ae35c6a8232af153acb0f366eafad59f48c4/portal-impl/src/com/liferay/portal/internal/servlet/MainServlet.java#L1119).
+
+So, whenever someone hits http://localhost:8080, the following requests are serviced by Tomcat:
+
+1. `/`: the originating external request
+2. `/web/guest`: internal request to which `/` is forwarded. Note that you can also request `/web/guest` from the browser and you get the same content.
+3. `/c/portal/layout?p_l_id=19&p_v_l_s_g_id=0`: this is internal too. The server maps the `/web/guest` request (which is a virtual path) to the configured layout, identified by a `p_l_id`. This is the request that truly generates the HTML content the user sees.
+4. `/o/js_loader_config?t=1600876854212`: this is a external request triggered by the HTML once the browser loads it. It returns the configuration of the AMD loader.
```

Comment:
Is it worth replacing the concrete timestamp?

> `/o/js_loader_config?t=TIMESTAMP` (where `TIMESTAMP` is an epoch timestamp in milliseconds such as `1600876854212`):

Bonus: can we link to a sample of what the AMD loader configuration looks like? (Could be in a gist).

Reply:
Mmm, I don't think we should replace the parameter values by labels because then we have to change almost all of them (`p_l_id`, `p_v_l_s_g_id`, ... aren't fixed either) and the URL will look weird and lose its usual appearance.

Regarding the sample loader config, it is out of the scope of this doc, I think... But if you insist, I don't mind creating the gist so that we can link it from other places too if necessary.

Reply:
But you have to insist. Otherwise I won't do it :monocle_face:

Reply:
>  I don't think we should replace the parameter values by labels because then we have to change almost all of them

I'd only do it for the first URL, just to show that it's a timestamp. I'd leave all the others unchanged in this doc, and I wouldn't both substituting things that are `id` because those are obviously ids.

But it's up to you.

As for the loader config, it just seemed like a pretty easy thing to do...

https://gist.github.com/wincent/bac7fec66df5a4b70c498b39fc91215b

```js
(function() {Liferay.EXPLAIN_RESOLUTIONS = false;
Liferay.EXPOSE_GLOBAL = false;
Liferay.LOG_LEVEL = 'warn';
Liferay.RESOLVE_PATH = "/o/js_resolve_modules";
Liferay.WAIT_TIMEOUT = 7000;
}());
```

Reply:
Added :+1:  (because you insisted). Thx for the gist.